src/null-safety/migrating-to-null-safety/index.md

```diff
@@ -0,0 +1,435 @@
+---
+title: Migrating to null safety
+description: How to move your existing Dart code to the world of null safety
+---
+
+**TODO: Null safety hasn't shipped yet and this doc still contains some TODOs
+for aspects of null safety (like the shipping stable version number) that aren't
+known yet. Fix those before landing this on master.**
+
+When we decided to add null safety to Dart, we wanted to get as much value out
+of it as possible. That meant designing a fully sound system that guarantees you
+will not see a null reference error. This lets our compilers optimize code
+knowing that `null` cannot flow into certain places. The challenge is that the
+world's existing Dart code is not designed for that level of strictness. How do
+we give users sound null safety without breaking all of the existing Dart code
+in the world?
+
+Our answer is that we let you *migrate* your code to null safety when *you* want
+to. This document explains how that works, and how to get your code into the new
+magical land of null safety.
+
+## The goal
+
+Our goal is that eventually the entire Dart ecosystem will move all of their
+Dart code to null safety. Null unsafe **legacy** Dart code is still supported
+for compatibility but is not a language we want users to be programming in
+forever. We think null safe Dart is a better language and want everyone to
+eventually be using it. At some point in the future, anyone surveying the Dart
+landscape should not realize that a null unsafe version of Dart ever existed.
+
+That's the place we want to reach. But we can't get there by breaking all of
+your code. It's not enough for us to settle *ourselves* on a new tropical island
+paradise. We have to build a bridge to the one you are all on and help you cross
+over.
+
+## Language versioning
```

Comment:
Should we move language versioning to a separate page? It's a general topic we'll need to refer to again? We could also link to such a page from places like https://dart.dev/null-safety#configure-the-sdk-version

Reply:
Eventually, yes. But at least right now, null safety is the only real reason a user would want to know about language versioning and I like that it keeps the doc self-contained.

Reply:
Is that true? For any release where we've shipped a new language feature since 2.6, I'd argue that they needed to use language versioning (we just didn't call it that).